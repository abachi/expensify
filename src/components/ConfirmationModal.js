import React from "react";
import Modal from "react-modal";

export const ConfirmationModal = (props) => {
  return (
    <Modal
      ariaHideApp={false}
      className="modal"
      isOpen={props.isOpen}
      onRequestClose={props.closeModal}
    >
      <div>
        <h3 className="modal__title">Do you want remove this expense?</h3>
        <div className="modal__body">
          <button
            name="cancel-btn"
            className="button button--secondary"
            onClick={props.onRequestClose}
          >
            Cancel
          </button>
          <button className="button button--danger" onClick={props.onRemove}>
            Sure
          </button>
        </div>
      </div>
    </Modal>
  );
};

export default ConfirmationModal;
