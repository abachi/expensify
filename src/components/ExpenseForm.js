import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? props.expense.amount : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    const isValid = /^\d+(\.\d{0,2})?$/.test(amount);
    if (!amount || isValid) {
      this.setState(() => ({ amount }));
    }
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onCalendarFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  handleSubmit = (e) => {
    e.preventDefault();
    const { description, note, amount, createdAt } = this.state;
    if (description.trim().length === 0 || !amount) {
      this.setState(() => ({
        error: "Please provide description and amount.",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: description.trim(),
        amount: parseFloat(amount, 10),
        note: note.trim(),
        createdAt: createdAt.valueOf(),
      });
    }
  };

  render() {
    const {
      description,
      note,
      amount,
      error,
      createdAt,
      calendarFocused,
    } = this.state;
    return (
      <form className="form" onSubmit={this.handleSubmit}>
        {error.length > 0 && <p className="form__error">{error}</p>}
        <input
          type="text"
          className="text-input"
          name="description"
          placeholder="Description"
          value={description}
          onChange={this.onDescriptionChange}
          autoFocus
        />
        <input
          type="text"
          className="text-input"
          name="amount"
          placeholder="Amount"
          value={amount}
          onChange={this.onAmountChange}
        />
        <textarea
          className="textarea"
          placeholder="Add a note (Optional)"
          value={note}
          name="note"
          onChange={this.onNoteChange}
        ></textarea>
        <SingleDatePicker
          date={createdAt}
          onDateChange={this.onDateChange}
          focused={calendarFocused}
          onFocusChange={this.onCalendarFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <div>
          <button className="button">Save Expense</button>
        </div>
      </form>
    );
  }
}
