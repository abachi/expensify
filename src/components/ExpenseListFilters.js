import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";

import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  state = {
    focusedInput: null,
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocus) => {
    this.setState(() => ({ calendarFocus }));
  };
  onTextSearchChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortByChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    }

    if (e.target.value === "date") {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search expenses"
              type="text"
              name="search"
              value={this.props.filters.text}
              onChange={this.onTextSearchChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              name="sortBy"
              onChange={this.onSortByChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.focusedInput}
              onFocusChange={this.onFocusChange}
              startDateId={"expensefy_0"}
              endDateId={"expensefy_9"}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

const mapStateToProps = (state) => ({
  filters: state.filters,
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
