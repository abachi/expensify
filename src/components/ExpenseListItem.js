import React from "react";
import { Link } from "react-router-dom";

export default (props) => (
  <div>
    <span>Amount: {props.expense.amount}</span>
    <span>Description: {props.expense.description}</span>
    <span>created at: {props.expense.createdAt}</span>
    <Link to={`edit/${props.expense.id}`}>Edit</Link>
  </div>
);
