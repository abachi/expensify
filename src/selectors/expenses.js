import moment from "moment";

export default (expenses, { text, startDate, endDate, sortBy }) => {
  return expenses
    .filter((expense) => {
      const textMatch =
        text.length === 0 ||
        expense.description.toLowerCase().includes(text.toLowerCase());
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;

      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt > b.createdAt ? -1 : 1;
      } else {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};
