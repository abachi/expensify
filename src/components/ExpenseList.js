import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";

export const ExpenseList = (props) => {
  const total = getExpensesTotal(props.expenses);
  return (
    <div>
      {props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        <div>
          <p>
            Viewing {props.expenses.length} expenses totaling{" "}
            {numeral(total).format("$0,0.00")}
          </p>
          {props.expenses.map((expense) => (
            <ExpenseListItem key={expense.id} expense={expense} />
          ))}
        </div>
      )}
    </div>
  );
};

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters),
});
export default connect(mapStateToProps)(ExpenseList);
