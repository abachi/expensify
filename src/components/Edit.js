import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class Edit extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div className="page">
        <div className="page-header">
          <div className="content-container">
            <h1>Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button--secondary" onClick={this.onRemove}>
            Remove Expense
          </button>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense)),
});
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((exp) => exp.id === props.match.params.id),
});
export default connect(mapStateToProps, mapDispatchToProps)(Edit);
