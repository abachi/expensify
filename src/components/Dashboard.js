import React from "react";
import ExpenseList from "./ExpenseList";
import ExpensesSummary from "./ExpensesSummary";
import ExpenseListFilters from "./ExpenseListFilters";
const Dashboard = () => (
  <div>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
export default Dashboard;
