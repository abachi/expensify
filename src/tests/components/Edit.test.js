import React from "react";
import { shallow } from "enzyme";
import { Edit } from "../../components/Edit";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

let editExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <Edit
      editExpense={editExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test("should render the Edit correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should submit and handle expense updates", () => {
  wrapper.find(ExpenseForm).prop("onSubmit")(expenses[0]);
  expect(editExpense).toHaveBeenCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenCalled();
});

test("should handle remove expense", () => {
  wrapper.find("button").prop("onClick")(expenses[0]);
  expect(startRemoveExpense).toHaveBeenCalledWith(expenses[0]);
  expect(history.push).toHaveBeenCalled();
});
