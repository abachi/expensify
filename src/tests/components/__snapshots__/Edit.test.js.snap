// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Edit correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1>
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 1500,
          "createdAt": 0,
          "description": "Rent",
          "id": "1",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
